src/messages.rs

```diff
@@ -15,178 +15,186 @@
+        
+        send_on_message.source = types::SourceAddress {
+            from_node : our_name.clone(),
+            from_group : Some(self.destination.dest.clone()),
+            reply_to : self.source.reply_to.clone(),
+            relayed_for : self.source.relayed_for.clone()
+        };
+        send_on_message.source = match self.source {
+              SourceAddress::RelayedForClient(_, b) => SourceAddress::RelayedForClient(our_name.clone(), b),
+              SourceAddress::RelayedForNode(_, b)   => SourceAddress::RelayedForNode(our_name.clone, b),
+              SourceAddress::Direct(a)              => SourceAddress::Direct(our_name.clone()),  
+        };
+
+        send_on_message.destination = match self.destination {
+            DestinationAddress::RelayToClient(_, b) => DestinationAddress::RelayToClient(destination, b),
+            DestinationAddress::RelayToNode(_, b)   => DestinationAddress::RelayToNode(destination, b),
+            DestinationAddress::Direct(_)           => DestinationAddress::Direct(destination),
+        };
+        send_on_message.authority = our_authority.clone();
+        send_on_message
+    }
+
+    /// This creates a new message for Action::Reply. It clones all the fields,
+    /// and then mutates the destination and source accordingly.
+    /// Authority is changed at this point as this method is called after
+    /// the interface has processed the message.
+    /// Note: this is not for XOR-forwarding; then the header is preserved!
+    pub fn create_reply(&self, our_name : &NameType, our_authority : &Authority)
+                        -> RoutingMessage {
+        // implicitly preserve all non-mutated fields.
+        // TODO(dirvine) Again why copy here instead of change in place?  :08/07/2015
+        let mut reply_message = self.clone();
+        reply_message.source  = match self.destination {
+            DestinationAddress::RelayToClient(_, b) => SourceAddress::RelayedForClient(our_name.clone(), b),
+            DestinationAddress::RelayToNode(_, b)   => SourceAddress::RelayedForNode(our_name.clone()), 
+            DestinationAddress::Direct(_)           => SourceAddress::Direct(our_name.clone()),
+        };
+        reply_message.destination = match self.source {
+            SourceAddress::RelayedForClient(a, b) => DestinationAddress::RelayToClient(a, b),
+            SourceAddress::RelayedForNode(a, b)   => DestinationAddress::RelayToNode(a, b),
+            SourceAddress::Direct(a)              => DestinationAddress::Direct(a),
+        };
+        reply_message.authority = our_authority.clone();
+        reply_message
     }
 }
+
+#[derive(PartialEq, Eq, Clone, Debug, RustcEncodable, RustcDecodable)]
+pub struct SignedRoutingMessage {
+    pub encoded_routing_message : Vec<u8>,
+    pub signature               : Signature
+}
+
+#[derive(PartialEq, Eq, Clone, Debug, RustcEncodable, RustcDecodable)]
+enum Message {
+Signed(SignedRoutingMessage),
+Unsigned(RoutingMessage)    
+}
```

Comment:
I did look at that, but the issue came down to multiple serialisation, I am not 100% of the efficiency, but we have to serialise the RoutingMessage anyway to sign, so I figured wrapping that was easiest. Ideally though I feel we need to get the DestinationAddress and SourceAddress easily parsable on their own so intermediate nodes do not need to parse everything all of the time.